lightning-c-bindings/README.md

```diff
@@ -0,0 +1,193 @@
+   take the relevant mapped type arguments.
+
+   Functions which return `&OpaqueRustType` and which return `OpaqueRustType` are both mapped to a
+   function returning an owned wrapper struct. The `_underlying_ref` flag (see above) will be set
+   to indicate that the pointed-to Rust object is owned or only a reference. Thus, when
+   implementing a function which Rust will call or calling a Rust function, you should check the
+   Rust documentation for the function to determine which behavior is expected.
+
+   Similarly, when a function takes an `Option<RustType>` as a parameter or a return value, the C
+   type is the same as if it took only `RustType`, with the `inner` field set to NULL to indicate
+   None. For example, `ChannelManager_create_channel` takes an `Option<LDKUserCOnfig>` not an
+   `LDKUserConfig`, but its definition is:
+   ```
+   MUST_USE_RES LDKCResult_NoneAPIErrorZ ChannelManager_create_channel(const LDKChannelManager *this_arg, ..., LDKUserConfig override_config);
+   ```
+
+ * Various containers (Tuples, Vecs, Results, etc) are mapped into C structs of the form
+   `LDKCContainerType_ContainerElementsZ`. Inner fields are often pointers, and in the case of
+   primitive types, these may be allocated in C using the system allocator. See [the Rust docs on
+   your platform's default System allocator](https://doc.rust-lang.org/std/alloc/struct.System.html)
+   for which allocator you must use.
+
+As the bindings are auto-generated, the best resource for documentation on them is the native Rust
+docs available via `cargo doc` or [docs.rs/lightning](https://docs.rs/lightning).
+
+The memory model is largely the Rust memory model and not a native C-like memory model. Thus,
+function parameters are largely only ever passed by reference or by move, with pass-by-copy
+semantics only applying to primitive types. However, because the underlying types are largely
+pointers, the same function signature may imply two different memory ownership semantics. Thus, you
+MUST read the Rust documentation while using the C bindings. For functions which take arguments
+where ownership is moved to the function scope, the corresponding `X_free` function MUST NOT be
+called on the object, whereas for all other objects, `X_free` MUST be used to free resources.
+
+LDK C++ Bindings
+================
+
+The C++ bindings available at include/lightningpp.hpp require extern "C" inclusion of lightning.h
+and rust_types.h first. They represent thin wrappers around the C types which provide a few
+C++-isms to make memory model correctness easier to achieve. They provide:
+ * automated destructors which call the relevant `X_free` C functions,
+ * move constructors both from C++ classes and the original C struct, with the original object
+   cleared to ensure destruction/`X_free` calls do not cause a double-free.
+ * Move semantics via the () operator, returning the original C struct and clearing the C++ object.
```

Comment:
Huh.. Not 100% but these lines make me think we should open an issue to bring the C demo to "feature parity" with the C++ demo, to make sure people using the C bindings have a decent example of everything and don't get "gotcha"'d by stuff the C++isms are achieving automatically, while C might have to be more explicit about it.

Reply:
#682.